plugins/dbus_announce.c

```diff
@@ -0,0 +1,144 @@
+#include "system.h"
+
+#include <dbus/dbus.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <rpm/rpmlog.h>
+#include <rpm/rpmts.h>
+#include <rpm/rpmdb.h>
+#include "lib/rpmplugin.h"
+
+struct dbus_announce_data {
+    DBusConnection * bus;
+    int old_dboffset;
+};
+
+static rpmRC dbus_announce_init(rpmPlugin plugin, rpmts ts)
+{
+    struct dbus_announce_data * state = rcalloc(1, sizeof(*state));
+    rpmPluginSetData(plugin, state);
+    return RPMRC_OK;
+}
+
+static void dbus_announce_close_bus(struct dbus_announce_data * state)
+{
+    if (state->bus) {
+	dbus_connection_close(state->bus);
+	dbus_connection_unref(state->bus);
+	state->bus = NULL;
+    }
+}
+
+static rpmRC open_dbus(rpmPlugin plugin, rpmts ts)
+{
+    DBusError err;
+    int rc = 0;
+    struct dbus_announce_data * state = rpmPluginGetData(plugin);
+
+    /* Already open */
+    if (state->bus)
+	return RPMRC_OK;
+
+    /* ...don't notify on test transactions */
+    if (rpmtsFlags(ts) & (RPMTRANS_FLAG_TEST|RPMTRANS_FLAG_BUILD_PROBS))
+	return RPMRC_OK;
+
+    /* ...don't notify on chroot transactions */
+    if (!rstreq(rpmtsRootDir(ts), "/"))
+	return RPMRC_OK;
+
+    dbus_error_init(&err);
+
+    state->bus = dbus_bus_get_private(DBUS_BUS_SYSTEM, &err);
+    if (dbus_error_is_set(&err)) {
+	goto err;
+    }
+    if (state->bus) {
+	rc = dbus_bus_request_name(state->bus, "org.rpm.announce",
+				   DBUS_NAME_FLAG_REPLACE_EXISTING , &err);
+    }
+    if (dbus_error_is_set(&err)) {
+	goto err;
+    }
+    if (DBUS_REQUEST_NAME_REPLY_PRIMARY_OWNER != rc) {
+	dbus_announce_close_bus(state);
+	goto err;
+    }
+    return RPMRC_OK;
+ err:
+    rpmlog(RPMLOG_WARNING,
+	   "dbus_announce plugin: Error connecting to dbus (%s)\n",
+	   err.message);
+    dbus_error_free(&err);
+    return RPMRC_OK;
+}
+
+static void dbus_announce_cleanup(rpmPlugin plugin)
+{
+    struct dbus_announce_data * state = rpmPluginGetData(plugin);
+    dbus_announce_close_bus(state);
+    free(state);
+}
+
+static rpmRC send_ts_message(rpmPlugin plugin,
+			     const char * name,
+			     rpmts ts,
+			     int res)
+{
+    struct dbus_announce_data * state = rpmPluginGetData(plugin);
+    DBusMessage* msg;
+    dbus_uint32_t serial = 0;
+    char * dbcookie;
+
+    if (!state->bus) {
+	return RPMRC_OK;
+    }
+
+    msg = dbus_message_new_signal("/org/rpm/Transaction", /* object name */
+				  "org.rpm.Transaction",  /* interface name */
+				  name);                  /* signal name */
+    if (NULL == msg) {
+	goto err;
+    }
+
+    dbcookie = rpmdbCookie(rpmtsGetRdb(ts));
+    if (!dbus_message_append_args(msg, DBUS_TYPE_STRING, &dbcookie)) {
+	goto err;
+    }
+    dbcookie = _free(dbcookie);
```

Comment:
dbcookie will leak if dbus_message_append_args() fails. The generally fail-safe strategy is to initialize to NULL and free at central exit point than try be clever with local allocations.

Reply:
Plugged the leak. Still not a text book example of the central exit point pattern...